Several fixes

Tieske opened the pull request:
3 commits for individual issues.

The 2nd and 3rd were issues also encountered by others (looking at the git graph)

The proper fix for the last one would be to dynamically generate it from the openresty directive. But I do not how to get that info.

Tieske committed:
fix(reuse) allow reuse of sockets

Luasocket doesn't allow reusing the same socket after it was closed.
This doesn't play nice with resty-http which allows this.

Tieske committed:
fix(tls) allow other tls versions

Tieske committed:
fix(flatten) handle numbers properly

thibaultcha commented on lib/resty/socket.lua:
```diff
@@ -56,12 +68,13 @@ do
       local params = {
         mode = 'client',
-        protocol = 'tlsv1',
+        protocol = 'any',
```
For this fix we'd want to introduce `options.protocol`, the patch should be identical to the one in lua-cassandra: https://github.com/thibaultcha/lua-cassandra/pull/141

Tieske replied:
shouldn't we leave this at `any` and then make the `options` below configurable? That would allow full control to the user, but leave room for flexibility, and be secure by excluding the bad ones as a default?

Tieske replied:
added

thibaultcha commented on lib/resty/socket.lua:
```diff
@@ -56,12 +68,13 @@ do
         verify = verify and 'peer' or 'none',
-        options = 'all'
+        -- can the following be dynamically generated from `lua_ssl_protocols`?
+        options = {"all", "no_sslv2", "no_sslv3", "no_tlsv1"}
```
These are good to have, nice (cc @jeremymv2 we should add these to lua-cassandra/pgmoon as well)

thibaultcha commented on lib/resty/socket.lua:
```diff
@@ -22,6 +24,16 @@ do
+        self._connected_before = true
+      end
+      return self.sock:connect(...)
+    end,
```
This patch has to somehow prevent calling `connect()` twice, or else it seems to me like doing so will simply replace the socket; what happens to the established connection and existing socket? Does the connection RST? If so, that's no good.

Tieske replied:
I tried initially, because the socket gave me a "closed" error. So I set a flag on the `close` and `keepalive` methods, and only renewed the socket when connect was called and that flag set. But that doesn't work when the remote closes the connection. And then we still get the "closed" error upon calling `connect` again. So this is the best I could come up with.

At worst it obfuscates the problem, if in the middle of a connection you call connect again, that should always be an error (resty and luasocket). So in this case the behaviour still does something weird, replacing the socket, which will cause other application errors probably and hence obfuscate the real issue. That said, this only happens in an already erroneous situation. While it fixes the good-path of reusing sockets after they are closed.

The only way around it is probably having a fail-safe way of detecting the socket to be closed

Tieske replied:
@thibaultcha did the patch below resolve this one? or do you have concerns left?

Tieske replied:
I rebased the PR, patch went in to the original commit, I'll resolve this, please have another look

Tieske commented on lib/resty/socket.lua:
```diff
@@ -22,6 +24,16 @@ do
   end
 
   proxy_mt = {
+    connect = function(self, ...)
+      if self._connected_before then
+        -- luasocket does not allow reusing sockets after being closed
+        local socket = require 'socket'
+        self.sock = socket.tcp()
+      else
+        self._connected_before = true
+      end
+      return self.sock:connect(...)
+    end,
```
Thinking about it, how about this, this should reduce the error margin:
```suggestion
    connect = function(self, ...)
      local ok, err = self.sock:connect(...)
      
      if self._connected_before and err == "closed" then
        -- luasocket does not allow reusing sockets after being closed
        local socket = require 'socket'
        self.sock = socket.tcp()
        ok, err = self.sock:connect(...)
      end
      
      self._connected_before = true
      return ok, err
    end,
```

Tieske committed:
fix(sslhandshake) support send_status_req

Tieske committed:
feat(ssl) make luasec fallback options configurable

Tieske committed:
chore(*) add .editorconfig file

Tieske commented:
@thibaultcha is anything else required?

Tieske committed:
feat(settimeouts) add compatibility function sock:settimeouts

Tieske commented:
added a new commit for `settimeouts` just to keep this branch/PR up to date with problems/fixes encountered elsewhere

Tieske committed:
feat(tls) add option to get the current config

This allows changing only a single option, by first getting the current
options, then modifying it, and setting it again.

Tieske commented:
closing this, created a fork at https://github.com/Tieske/lua-resty-luasocket
